### PR TITLE
Variable Backup Files Purger

### PR DESCRIPTION
**Pull request by TheAbsolutionism**
### Description
This PR allows Skript to purge backup files, if number of files is >= `required` and will delete files starting from the oldest until number of files left = `remain` 

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/resources/config.sk
Comment (Efnilite) [inline src/main/resources/config.sk:1]: why did you change a bunch of stuff unrelated to your changes?

Comment (sovdeeth) [reply src/main/resources/config.sk:1]: Looks like a bad merge to me

Comment (TheAbsolutionism) [reply src/main/resources/config.sk:1]: It was unintentional. I think I overwrote from my outdated `config.sk` on my localhost for some reason. I can fix

File: src/main/java/ch/njol/skript/util/FileUtils.java [hunk 1]
```diff
@@ -60,7 +63,21 @@ public static String getBackupSuffix() {
...
+		if (!backupDir.exists() || !backupDir.isDirectory())
+			throw new IOException("Backup directory not found");
+		final ArrayList<File> files = new ArrayList<File>(Arrays.asList(backupDir.listFiles()));
+		if (files == null || files.size() < required)
+			return;
+		files.sort(Comparator.comparingLong(File::lastModified));
+		final int iterations = files.size() - remain;
+		for (int i = 0; i < iterations; i++) {
+			files.get(i).delete();
+		};
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/util/FileUtils.java:78]:
```suggestion
		}
```

File: src/main/java/ch/njol/skript/variables/VariablesStorage.java [hunk 2]
```diff
@@ -224,9 +224,26 @@ public final boolean load(SectionNode sectionNode) {
 			// Set the backup interval, if present & enabled
 			if (!"0".equals(getValue(sectionNode, "backup interval"))) {
 				Timespan backupInterval = getValue(sectionNode, "backup interval", Timespan.class);
-
+				boolean purge = getValue(sectionNode, "backup purge", Boolean.class);
+				final Integer required = getValue(sectionNode, "purge required", Integer.class);
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/variables/VariablesStorage.java:228]: can't you use primitives here like on previous line?

File: src/main/java/ch/njol/skript/variables/VariablesStorage.java [hunk 3]
```diff
@@ -307,11 +324,10 @@ public final boolean load(SectionNode sectionNode) {
 	 *
 	 * @param backupInterval the backup interval.
 	 */
-	public void startBackupTask(Timespan backupInterval) {
+	public void startBackupTask(Timespan backupInterval, Boolean purge, Integer required, Integer remain) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/variables/VariablesStorage.java:327]: should also probably use primitives here

File: src/main/resources/config.sk [hunk 4]
```diff
@@ -338,6 +338,11 @@ databases:
 
 		backup interval: 2 hours
 
+		backup purge: false #set to "true" to enable the automatic purging feature
```
Comment (Efnilite) [inline src/main/resources/config.sk:341]: maybe explain what purging does exactly on prior lines instead of on the same line

File: src/main/resources/config.sk [hunk 5]
```diff
@@ -338,6 +338,11 @@ databases:
...
+		purge required: 50 #Number of backup files required in order to start the purge process
+		purge remain: 20 #Number of files to be left remaining when purge starts
```
Comment (Efnilite) [inline src/main/resources/config.sk:343]: imo it's easier to explain if the config option is how many files to purge instead of the number of files are left over after purging

File: src/main/java/ch/njol/skript/variables/VariablesStorage.java (hunk 2 above)
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/variables/VariablesStorage.java:228]: @Efnilite I used Integer instead of int solely on the off chance an error occurs or invalid input from a user. Which allows me to check if it's null in the following checks. That way the purge feature can be disabled. If you think I should still go ahead and change it to int I can.

File: src/main/java/ch/njol/skript/variables/VariablesStorage.java (hunk 3 above)
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/variables/VariablesStorage.java:327]: I can change it to int within the method, since I am not doing any checks here

File: src/main/resources/config.sk [hunk 6]
```diff
@@ -289,6 +289,12 @@ databases:
 		# Variables are saved constantly no matter what is set here, thus a server crash will never make you loose any variables.
 		# Set this to 0 to disable this feature.
 
+		backups to keep: -1
+		# if enabled, will check the specified number of backup files in the backup directory each time a backup file is created
+		# If the number of backup files >= the number of 'backups to keep', it will delete files until number of files is equal to `backups to keep`
+		# If disabled, no backup files will be deleted
+		# By default, -1 equals disabled, set this to any number above 0 to enabled. Can not be set to 0
```
Comment (sovdeeth) [inline src/main/resources/config.sk:296]: Setting to 0 could disable backups too?

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/resources/config.sk (hunk 6 above)
Comment (TheAbsolutionism) [reply src/main/resources/config.sk:296]: @sovdeeth would you like for me to add that functionality? Where if they put 0, it doesn’t start the backup task?

Comment (sovdeeth) [reply src/main/resources/config.sk:296]: yes but i would like opinions

Comment (EquipableMC) [reply src/main/resources/config.sk:296]: I honestly feel like 0 should also disable it. Skript users sometimes don't know how configs work and would expect "0" to work and not realize -1 also works.

**Comment by TheAbsolutionism** [comment]
Feature does work as intended, including being disabled, only delete a certain amount of files, and delete all files + not starting backup task

**Comment by EquipableMC** [comment]
> Feature does work as intended, including: `-1` being disabled `# > 0` Delete oldest files until equals `#` `0` Delete all files + not starting backup task

How’s the performance? Any lag spikes, etc?

**Comment by TheAbsolutionism** [comment]
> How’s the performance? Any lag spikes, etc?
I didnt see/witness any lag spikes, then again, it was tested on a localhost

**Comment by Pikachu920** [comment]
can we add tests? i know its not gonna be super simple, but this seems like a feature which we definitely want to be sure works as intended

**Comment by TheLimeGlass** [comment]
I want to add that backups get created even when there are no variables in the database. CSV
Fuse has an issue about it.

**Comment by TheAbsolutionism** [comment]
> can we add tests?

I'm honestly not sure how to go about doing that. I've never done tests before, and from what I've seen, the tests are in Skript rather than Java. Plus, the main aspect relies on the setting within the config.

> I want to add that backups get created even when there are no variables in the database

If you want, I can add within `startBackupTask` if the bytes of the variables.csv file is equal to 0, then it wont create a backup.
Only problem is, if all global vars get deleted, you will still have the commented info at the top, which varies of how many bytes based on the characters.

**Comment by Fusezion** [comment]
> I want to add that backups get created even when there are no variables in the database. CSV Fuse has an issue about it.

The issue spoken of and this pr I believe should be left as standalone relations. The code iirc already is suppose to prevent it the issue lies within ram variables counting towards a saved variable. Which until a proper system for ram vars is implemented there's little safety methods possible. #6609 is a much better starting point for that issue until further efforts are put in

**Comment by Pikachu920** [comment]
> I'm honestly not sure how to go about doing that. I've never done tests before, and from what I've seen, the tests are in Skript rather than Java. Plus, the main aspect relies on the setting within the config.
> 
> >

you can write tests in Java using the junit testing system. I can give it a shot if you dont mind

File: src/main/java/ch/njol/skript/util/FileUtils.java [hunk 7]
```diff
@@ -60,7 +63,24 @@ public static String getBackupSuffix() {
 			return "" + backupFormat.format(System.currentTimeMillis());
 		}
 	}
-	
+
+	public static void backupPurge(final File f, int tokeep) throws IOException {
+		if (tokeep < 0)
+			throw new IOException("Called with invalid input");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/FileUtils.java:69]: IllegalArgumentException (or use Preconditions) Also say why it's invalid.

File: src/main/java/ch/njol/skript/util/FileUtils.java [hunk 8]
```diff
@@ -60,7 +63,24 @@ public static String getBackupSuffix() {
 			return "" + backupFormat.format(System.currentTimeMillis());
 		}
 	}
-	
+
+	public static void backupPurge(final File f, int tokeep) throws IOException {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/FileUtils.java:67]:
```suggestion
	public static void backupPurge(File f, int tokeep) throws IOException {
```

File: src/main/java/ch/njol/skript/util/FileUtils.java [hunk 9]
```diff
@@ -60,7 +63,24 @@ public static String getBackupSuffix() {
...
+		final File backupDir = new File(f.getParentFile(), "backups" + File.separator);
+		if (!backupDir.exists() || !backupDir.isDirectory())
+			throw new IOException("Backup directory not found");
+		final ArrayList<File> files = new ArrayList<File>(Arrays.asList(backupDir.listFiles()));
+		if (files == null || files.size() <= tokeep)
+			return;
+		if (tokeep > 0)
+			files.sort(Comparator.comparingLong(File::lastModified));
+		final int Iterations = files.size() - tokeep;
+		for (int i = 0; i < Iterations; i++) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/FileUtils.java:79]:
```suggestion
		File backupDir = new File(f.getParentFile(), "backups" + File.separator);
		if (!backupDir.exists() || !backupDir.isDirectory())
			throw new IOException("Backup directory not found");
		ArrayList<File> files = new ArrayList<File>(Arrays.asList(backupDir.listFiles()));
		if (files == null || files.size() <= tokeep)
			return;
		if (tokeep > 0)
			files.sort(Comparator.comparingLong(File::lastModified));
		int Iterations = files.size() - tokeep;
		for (int i = 0; i < Iterations; i++) {
```

Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/FileUtils.java:79]:
```suggestion
		final int numberToRemove = files.size() - tokeep;
		for (int i = 0; i < numberToRemove; i++) {
```

File: src/main/java/ch/njol/skript/variables/VariablesStorage.java [hunk 10]
```diff
@@ -224,9 +225,24 @@ public final boolean load(SectionNode sectionNode) {
...
+					if (tokeep == 0) {
+						proceed = false;
+					} else if (tokeep >= 1) {
+						purge = true;
+					}
+					if (proceed) {
+                        startBackupTask(backupInterval, purge, tokeep);
+                    } else {
+						try {
+							FileUtils.backupPurge(file, tokeep);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/variables/VariablesStorage.java:241]: The naming schema here's a bit confusing, with the various meanings of purge, the ambiguity of proceed, etc. This should be improved for clarity.

File: src/main/java/ch/njol/skript/variables/VariablesStorage.java [hunk 11]
```diff
@@ -307,11 +323,10 @@ public final boolean load(SectionNode sectionNode) {
 	 *
 	 * @param backupInterval the backup interval.
 	 */
-	public void startBackupTask(Timespan backupInterval) {
+	public void startBackupTask(Timespan backupInterval, boolean purge, int tokeep) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/variables/VariablesStorage.java:326]: Like above, purge isn't a great name. Something like "removeExtra" would be better. Also capitalize toKeep

File: src/main/resources/config.sk [hunk 12]
```diff
@@ -289,6 +289,14 @@ databases:
 		# Variables are saved constantly no matter what is set here, thus a server crash will never make you loose any variables.
 		# Set this to 0 to disable this feature.
 
+		backups to keep: -1
+		# if enabled, will check the amount of backup files in the backup directory each time a backup file is created
+		# If the number of backup files > the number of 'backups to keep', it will delete files until number of files is equal to `backups to keep`
+		# If disabled, no backup files will be deleted
+		# By default, -1 equals disabled, set this to any number above 0 to enable it.
+		# By setting to 0, will disable any new variable backups from being created
+		# WARNING: Setting to 0 will delete all files located in the backup directory upon plugin start/reload
```
Comment (sovdeeth) [inline src/main/resources/config.sk:298]:
```suggestion
		# if set to a value greater than -1, Skript will ensure only that many backups are kept. 
		# Backups are removed from oldest to newest.
		# If disabled (set to -1), no backup files will be deleted.
		# WARNING: Setting to 0 will delete all files located in the backup directory upon plugin start/reload
```

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/util/FileUtils.java (hunk 7 above)
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/util/FileUtils.java]: @sovdeeth Through Skript itself, it wont call this method if the `toKeep` is less than 0, This check is purely just incase anyone calls it through skript-reflect for example.

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/util/FileUtils.java]: That's why you should use IllegalArgumentException. It's an illegal argument, not an io error

Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/util/FileUtils.java]: Yea, I added it

File: src/main/java/ch/njol/skript/util/FileUtils.java [hunk 13]
```diff
@@ -60,7 +63,24 @@ public static String getBackupSuffix() {
 			return "" + backupFormat.format(System.currentTimeMillis());
 		}
 	}
-	
+
+	public static void backupPurge(File f, int toKeep) throws IOException, IllegalArgumentException {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/FileUtils.java:67]: what's f for? f doesn't tell me anything. give it a better name
also this method should have javadocs

**Review by sovdeeth — CHANGES_REQUESTED** [review]
one last thing

**Review by sovdeeth — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/util/FileUtils.java [hunk 14]
```diff
@@ -60,7 +63,32 @@ public static String getBackupSuffix() {
...
+
+	/**
+	 * Deletes files in backup directory to meet desired target, starting from oldest to newest
+	 *
+	 * @param csvFile Variable file in order to get 'backups' directory
+	 * @param toKeep Integer of how many files are to be left remaining
+	 * @throws IOException If 'backups' directory is not found
+	 * @throws IllegalArgumentException If 'toKeep' parameter is less than 0
+	 */
+	public static void backupPurge(File csvFile, int toKeep) throws IOException, IllegalArgumentException {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/util/FileUtils.java:75]: Since this is only FlatFile CSV related, can this go within the FlatFileStorage class?

File: src/main/java/ch/njol/skript/variables/VariablesStorage.java [hunk 15]
```diff
@@ -224,9 +225,24 @@ public final boolean load(SectionNode sectionNode) {
...
+					}
+					if (startBackup) {
+                        startBackupTask(backupInterval, removeBackups, toKeep);
+                    } else {
+						try {
+							FileUtils.backupPurge(file, toKeep);
+						} catch (IOException e) {
+							Skript.error("Variables backup wipe failed: " + e.getLocalizedMessage());
+						}
+					}
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/variables/VariablesStorage.java:245]: Other databases can require files aswell but not all databases use CSV extension. Can this go within the FlatFileStorage class? You can use the `load_i` method within the FlatFileStorage to get the same access to the ConfigurationSection

File: src/main/resources/config.sk [hunk 16]
```diff
@@ -325,6 +331,8 @@ databases:
 		monitor changes: false
 		monitor interval: 20 seconds
 
+		backups to keep: -1
```
Comment (TheLimeGlass) [inline src/main/resources/config.sk:334]: You can use FlatFileStorage#load_i method to avoid requiring this on every database.

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

**Review by Efnilite — APPROVED** [review]
